### PR TITLE
Add Minitron pruning example for NeMo 2.0

### PR DESCRIPTION
**Pull request by kevalmorabia97**
# What does this PR do ?

Add Minitron pruning example for NeMo 2.0

**Collection**: N/A

# Changelog 
- Add Minitron pruning example for NeMo 2.0

# Usage
* You can potentially add a usage example below

```python
torchrun --nproc_per_node 8 gpt_prune.py \
        --devices 8 \
        --tp_size 1 \
        --pp_size 8 \
        --restore_path <path/to/llama3.1-8b-nemo2> \
        --seq_length 8192 \
        --data_paths 30 path/to/dataset_1_prefix 70 path/to/dataset_2_prefix \
        --prune_ffn_hidden_size 9216 \
        --prune_hidden_size 3072 \
        --prune_num_attention_heads 32 \
        --prune_num_query_groups 8 \
        --save_path llama3.1-8b-width-pruned
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: scripts/llm/prune.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+
+def save_pruned_model(model: llm.GPTModel, trainer: nl.Trainer, save_path: str):
+    """Save pruned model nemo checkpoint."""
+    logging.info(f"Saving pruned model to {save_path}...")
+    if hasattr(trainer.model, "__io__") and hasattr(trainer.model.tokenizer, "__io__"):
+        trainer.model.__io__.tokenizer = trainer.model.tokenizer.__io__
+        # Make sure pruned hparams are saved
+        for k in SUPPORTED_PRUNING_HPARAMS | {"kv_channels"}:
+            setattr(trainer.model.__io__.config, k, getattr(model.config, k))
```
Comment (AAnoosheh) [inline scripts/llm/prune.py:176]: Is this different than `model.__io__.config = model.config.__io__` ?

File: scripts/llm/prune.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+    weight_path = ckpt_to_weights_subdir(save_path, is_saving=True)
+    weight_path.mkdir(parents=True, exist_ok=True)
+    dist_checkpointing.save(trainer.strategy.megatron_parallel.sharded_state_dict(), weight_path)
```
[the remark this answers is not in the document]
Comment (kevalmorabia97) [reply scripts/llm/prune.py:187]: If I do that then there is a missing `.metadata` and `metadata.json` file in `save_path/weights` so when I try to load pruned checkpoint, I get `megatron.core.dist_checkpointing.core.CheckpointingException: <save_path>/weights is not a distributed checkpoint`

Did you see this issue?

File: scripts/llm/prune.py (hunk 1 above)
Comment (kevalmorabia97) [reply scripts/llm/prune.py:183]: If I do that then somehow the updated pruned values for hidden_size, ffn, attention, etc. done show up in `save_path/context/model.yaml`

Comment (AAnoosheh) [reply scripts/llm/prune.py:185]: Interesting... I'll need to figure out how this is working. Good for now though

File: nemo/lightning/fabric/plugins.py [hunk 3]
```diff
@@ -150,8 +150,9 @@ def convert_module(self, module: nn.Module) -> nn.Module:
             config.fp16 = self.dtype_config.fp16
             config.bf16 = self.dtype_config.bf16
             # Avoid rewrapping the module if it's already of type Float16Module
-            if hasattr(module, 'module') and not isinstance(module.module, Float16Module):
-                module.module = Float16Module(config, module.module)
+            if hasattr(module, 'module'):
```
Comment (kevalmorabia97) [inline nemo/lightning/fabric/plugins.py:153]: @farhadrgh @ananthsub fyi I had to add this fix after the recent change in https://github.com/NVIDIA/NeMo/pull/12141 to get this to work (`llm.validate` on pre-initialized model)

**Review by yashaswikarnati — APPROVED** [review]

**Review by ko3n1g — APPROVED** [review]
